File: lib/node/vssh-app-fs/ops.js
'use strict';

// As operações de filesystem, sem nada de transporte e sem nada de app específico.
//
// Proveniência: a semântica nasceu espelhando o processo main do Electron do Logseq, que é uma
// implementação de referência madura de um protocolo de fs para editor
// (`src/electron/electron/handler.cljs`, `deps/common/src/logseq/common/graph.cljs`). Isso vale
// como origem, não como política: as listas de extensão e de ignore que eram dele saíram para
// `presets/logseq.js` na promoção para o toolkit. Onde há divergência deliberada em relação
// àquela referência, está marcado com "DIVERGE:".

const fs = require('node:fs');
const fsp = require('node:fs/promises');
const path = require('node:path');

const { FsError, ENOENT, EINVAL, resolveInRoot, resolveRoot, notFound } = require('./paths');

// Extensões cujo conteúdo é lido e devolvido em open-dir/get-files.
//
// Este default é genérico de propósito: texto que qualquer app provavelmente quer ler inteiro.
// O que era específico do Logseq (a `allowed-formats` de logseq/common/graph.cljs) mora agora em
// `presets/logseq.js` — a lib não deve ter política de nenhum app embutida.
const DEFAULT_CONTENT_EXTENSIONS = [
  'txt', 'md', 'markdown', 'json', 'csv', 'yml', 'yaml', 'html', 'css', 'js', 'xml',
];

// Ignorados por padrão: só o que é cruft em qualquer projeto. Regras de app entram por config.
const DEFAULT_IGNORE = {
  prefixes: [],
  exact: [],
  dirNames: ['node_modules', '.git'],
  suffixes: ['.DS_Store'],
  hidden: true, // qualquer componente de caminho começando com "." (mas não "..")
};

// Onde `unlink` deposita o que remove. Relativo à raiz.
const DEFAULT_RECYCLE_DIR = '.recycle';

const DEFAULT_MAX_CONTENT_BYTES = 16 * 1024 * 1024;

function extOf(p) {
  const ext = path.extname(p);
  return ext.startsWith('.') ? ext.slice(1).toLowerCase() : ext.toLowerCase();
}

function hasHiddenComponent(rel) {
  return rel.split('/').some((seg) => seg.length > 1 && seg.startsWith('.') && seg !== '..');
}

function makeIsIgnored(ignore) {
  const cfg = { ...DEFAULT_IGNORE, ...ignore };
  isIgnored.config = cfg;   // o walk precisa da regra de ocultos separada do resto
  return isIgnored;

  function isIgnored(rel) {
    if (!rel) return false;
    const p = rel.split(path.sep).join('/');
    if (cfg.hidden && hasHiddenComponent(p)) return true;
    if (cfg.prefixes.some((prefix) => p === prefix || p.startsWith(prefix + '/'))) return true;
    if (cfg.exact.includes(p)) return true;
    if (cfg.dirNames.some((name) => p.split('/').slice(0, -1).includes(name))) return true;
    if (cfg.suffixes.some((suffix) => p.endsWith(suffix))) return true;
    return false;
  }
}

function statToInfo(stat) {
  return {
    type: stat.isDirectory() ? 'directory' : 'file',
    size: stat.size,
    // epoch ms — é o que `->db-files` põe em :file/last-modified-at.
    mtime: Math.round(stat.mtimeMs),
  };
}

/**
 * @param {object} config
 * @param {string} config.root                  raiz do grafo (criada se não existir)
 * @param {string[]} [config.contentExtensions] extensões cujo conteúdo entra em open-dir/get-files
 * @param {object} [config.ignore]              override das regras de ignore (ver DEFAULT_IGNORE)
 * @param {string} [config.recycleDir]          destino relativo do unlink (não apaga de verdade)
 * @param {number} [config.maxContentBytes]     acima disto o arquivo é listado sem conteúdo
 * @param {(event: object) => void} [config.onWarn] canal de avisos (o app decide se loga)
 */
function createAppFs(config = {}) {
  const realRoot = resolveRoot(config.root);
  const contentExtensions = new Set(
    (config.contentExtensions || DEFAULT_CONTENT_EXTENSIONS).map((e) => e.toLowerCase()),
  );
  const isIgnored = makeIsIgnored(config.ignore);
  const hiddenIgnored = isIgnored.config.hidden;
  const recycleDir = config.recycleDir || DEFAULT_RECYCLE_DIR;
  const maxContentBytes = config.maxContentBytes ?? DEFAULT_MAX_CONTENT_BYTES;
  const onWarn = config.onWarn || (() => {});

  const resolve = (input) => resolveInRoot(realRoot, input);

  // Caminhamento recursivo: descarta symlinks (um link para fora da raiz não vaza nem por
  // leitura), devolve só arquivos, caminhos absolutos.
  //
  // Ocultos NÃO são mais descartados aqui de forma incondicional. Antes eram, e isso fazia a
  // config mentir: `ignore.hidden = false` não tinha efeito nenhum, porque o walk já havia
  // derrubado os ocultos antes de `isIgnored` rodar. Agora quem decide é sempre a mesma regra.
  //
  // `applyIgnore` é opt-in porque as duas chamadoras querem coisas diferentes, e isso é
  // deliberado, não um descuido: `collectFiles` quer a visão filtrada (é o carregamento do
  // conteúdo), e `readdir` quer a visão crua — o Logseq a usa justamente para alcançar
  // `logseq/version-files`, que está na lista de ignore dele. Filtrar lá esconderia do consumidor
  // exatamente o que ele foi buscar.
  async function walk(startAbs, { applyIgnore = false } = {}) {
    const out = [];
    const queue = [startAbs];
    while (queue.length) {
      const dir = queue.pop();
      let entries;
      try {
        entries = await fsp.readdir(dir, { withFileTypes: true });
      } catch (err) {
        if (err.code === 'ENOENT') continue;
        throw err;
      }
      for (const entry of entries) {
        if (entry.isSymbolicLink()) continue;
        const abs = path.join(dir, entry.name);
        const rel = path.relative(realRoot, abs).split(path.sep).join('/');
        // Ocultos: sempre pela config, nunca hardcoded — é o que faz `hidden: false` funcionar.
        if (hiddenIgnored && hasHiddenComponent(rel)) continue;
        if (applyIgnore && isIgnored(rel)) continue;
        if (entry.isDirectory()) queue.push(abs);
        else out.push(abs);
      }
    }
    out.sort();
    return out;
  }

  // DIVERGE: o upstream lê qualquer tamanho. Um arquivo gigante na allowlist derrubaria o backend
  // na abertura do grafo.
  //
  // Ele é OMITIDO da listagem, não listado com conteúdo vazio: um arquivo que o Logseq não conhece
  // fica intocado, enquanto um arquivo listado como vazio é uma página vazia aos olhos dele — e o
  // primeiro save por cima apaga o conteúdo real no disco.
  function tooLarge(stat) {
    return stat.size > maxContentBytes;
  }

  // = `get-files` de common/graph.cljs: só as extensões da allowlist, ignorados fora, com conteúdo.
  async function collectFiles(dirAbs) {
    const all = await walk(dirAbs, { applyIgnore: true });
    const files = [];
    for (const abs of all) {
      if (!contentExtensions.has(extOf(abs))) continue;
      let stat;
      try {
        stat = await fsp.stat(abs);
      } catch (err) {
        if (err.code === 'ENOENT') continue; // corrida com escrita externa: só ignora
        throw err;
      }
      if (tooLarge(stat)) {
        onWarn({ event: 'file-skipped', reason: 'too-large', path: abs, size: stat.size, limit: maxContentBytes });
        continue;
      }
      const info = statToInfo(stat);
      files.push({
        path: abs,
        content: await fsp.readFile(abs, 'utf8'),
        size: info.size,
        mtime: info.mtime,
        type: info.type,
      });
    }
    return files;
  }

  async function openDir(input) {
    const { abs } = resolve(input || realRoot);
    const stat = await fsp.stat(abs).catch(() => null);
    if (!stat) throw notFound(input || realRoot);
    if (!stat.isDirectory()) throw new FsError(EINVAL, `não é diretório: ${abs}`);
    return { path: abs, files: await collectFiles(abs) };
  }

  return {
    root: realRoot,

    async stat(input) {
      const { abs } = resolve(input);
      try {
        return statToInfo(await fsp.stat(abs));
      } catch (err) {
        if (err.code === 'ENOENT') throw notFound(input);
        throw err;
      }
    },

    // "Existe?" é pergunta booleana e merece resposta booleana. `stat` responde com erro quando o
    // caminho não existe, e isso está certo — é a resposta correta para "me dê os metadados". Mas
    // quem só quer saber se existe acaba usando o erro como fluxo de controle, e aí toda sondagem
    // de rotina vira ruído de erro nos dois lados. Caminho fora da raiz continua sendo EACCES:
    // se existe ou não fora da raiz não é resposta que esta lib deva dar.
    async exists(input) {
      const { abs } = resolve(input);
      return { exists: (await fsp.stat(abs).catch(() => null)) !== null };
    },

    async readFile(input) {
      const { abs } = resolve(input);
      try {
        return { content: await fsp.readFile(abs, 'utf8') };
      } catch (err) {
        if (err.code === 'ENOENT') throw notFound(input);
        throw err;
      }
    },

    // `content` string (utf8) ou Buffer/Uint8Array (asset colado no editor).
    async writeFile(input, content) {
      const { abs } = resolve(input);

      // Gravar por cima de um diretório nunca é o que o chamador quis: significa que o caminho
      // chegou errado (vazio, resolvendo para a raiz do grafo, ou apontando para uma pasta). Sem
      // esta checagem o `fsp.writeFile` lança EISDIR, que o transporte não sabia classificar e
      // devolvia 500 — um pedido inválido do cliente aparecendo como defeito do servidor.
      const existing = await fsp.stat(abs).catch(() => null);
      if (existing && existing.isDirectory()) {
        throw new FsError(EINVAL, `destino de escrita é um diretório: ${abs}`);
      }

      // Conteúdo que não é texto nem bytes faria `Buffer.from` lançar um TypeError sem errno, que
      // também acabaria em 500. É entrada inválida, então é 400.
      const isBytes = content instanceof Uint8Array || ArrayBuffer.isView(content);
      if (typeof content !== 'string' && !isBytes) {
        throw new FsError(EINVAL, `conteúdo deve ser texto ou bytes, veio ${typeof content}`);
      }

      await fsp.mkdir(path.dirname(abs), { recursive: true });
      const data = typeof content === 'string' ? content : Buffer.from(content);
      await fsp.writeFile(abs, data);
      const stat = await fsp.stat(abs);
      const info = statToInfo(stat);
      return { path: abs, size: info.size, mtime: info.mtime };
    },

    async mkdir(input) {
      const { abs } = resolve(input);
      try {
        await fsp.mkdir(abs);
      } catch (err) {
        // O caller do Logseq trata EEXIST como sucesso (fs/node.cljs faz exatamente isso).
        if (err.code !== 'EEXIST') throw err;
      }
      return {};
    },

    async mkdirRecur(input) {
      const { abs } = resolve(input);
      await fsp.mkdir(abs, { recursive: true });
      return {};
    },

    // Enumerador cru: todos os arquivos, recursivo, caminhos absolutos, sem filtro de extensão.
    //
    // Por padrão NÃO aplica a `ignore` — e isso é deliberado, não uma divergência esquecida em
    // relação a open-dir/get-files. Quem chama readdir costuma estar atrás justamente do que a
    // ignore esconde (o Logseq a usa para alcançar `logseq/version-files`, que está na ignore
    // dele). Passe `{ applyIgnore: true }` para a visão filtrada. A regra de ocultos vale nos
    // dois modos, porque vem da config.
    async readdir(input, { applyIgnore = false } = {}) {
      const { abs } = resolve(input || realRoot);
      return walk(abs, { applyIgnore });
    },

    // Por padrão não apaga: move para o recycle dir com o caminho relativo achatado
    // (`replace "/" "_"`), para um "desfazer exclusão" continuar possível.
    //
    // `{ recycle: false }` apaga de verdade — para app que já tem lixeira própria, ou para o que
    // é descartável por natureza (cache, temporário) e não deve inchar o recycle dir.
    async unlink(input, { recycle = true } = {}) {
      const { abs, rel } = resolve(input);
      if (!recycle) {
        try {
          await fsp.unlink(abs);
        } catch (err) {
          if (err.code === 'ENOENT') throw notFound(input);
          throw err;
        }
        return { recycled: null };
      }
      const { abs: recycleAbs } = resolve(recycleDir);
      await fsp.mkdir(recycleAbs, { recursive: true });
      const flat = rel.split(path.sep).join('/').replace(/\//g, '_');
      const dest = path.join(recycleAbs, flat);
      try {
        await fsp.rename(abs, dest);
      } catch (err) {
        if (err.code === 'ENOENT') throw notFound(input);
        if (err.code === 'EXDEV') {
          // recycle em outro device (bind mount): copia e remove.
          await fsp.copyFile(abs, dest);
          await fsp.unlink(abs);
        } else {
          throw err;
        }
      }
      return { recycled: dest };
    },

    async rename(oldInput, newInput) {
      const { abs: oldAbs } = resolve(oldInput);
      const { abs: newAbs } = resolve(newInput);
      await fsp.mkdir(path.dirname(newAbs), { recursive: true });
      try {
        await fsp.rename(oldAbs, newAbs);
      } catch (err) {
        if (err.code === 'ENOENT') throw notFound(oldInput);
        throw err;
      }
      return {};
    },

    // Sobrescreve sem confirmação — o protocolo pede isso explicitamente ("copy with overwrite").
    async copy(oldInput, newInput) {
      const { abs: oldAbs } = resolve(oldInput);
      const { abs: newAbs } = resolve(newInput);
      await fsp.mkdir(path.dirname(newAbs), { recursive: true });
      try {
        await fsp.copyFile(oldAbs, newAbs);
      } catch (err) {
        if (err.code === 'ENOENT') throw notFound(oldInput);
        throw err;
      }
      return {};
    },

    // `{path, files}` — mesma forma que :openDir/:getFiles devolvem no Electron.
    openDir,

    // O protocolo distingue os dois (`open-dir` abre um grafo novo, `get-files` recarrega), mas a
    // leitura é a mesma; quem separa os dois casos é o frontend.
    getFiles: openDir,

    // Para o transporte servir binário (imagem, PDF) sem passar por JSON.
    async openRead(input) {
      const { abs } = resolve(input);
      const stat = await fsp.stat(abs).catch(() => null);
      if (!stat || stat.isDirectory()) throw notFound(input);
      const info = statToInfo(stat);
      return {
        path: abs,
        size: info.size,
        mtime: info.mtime,
        stream: (opts) => fs.createReadStream(abs, opts),
      };
    },
  };
}

module.exports = {
  createAppFs,
  DEFAULT_CONTENT_EXTENSIONS,
  DEFAULT_IGNORE,
  DEFAULT_RECYCLE_DIR,
  DEFAULT_MAX_CONTENT_BYTES,
  FsError,
  ENOENT,
  EINVAL,
};

File: lib/python/vssh_app_toolkit/fs/ops.py
"""As operações do filesystem privado, sem transporte.

O par do `lib/node/vssh-app-fs/ops.js`. Cada decisão aqui tem uma razão que já custou caro uma vez —
elas estão nos comentários, e não no histórico.
"""

from __future__ import annotations

import os
import shutil

from .paths import EINVAL, ErroFs, nao_encontrado, resolver_na_raiz, resolver_raiz

__all__ = ["criar_fs_do_app", "EXTENSOES_DE_CONTEUDO_PADRAO", "IGNORE_PADRAO",
           "DIR_RECICLAGEM_PADRAO", "MAX_BYTES_DE_CONTEUDO_PADRAO"]

EXTENSOES_DE_CONTEUDO_PADRAO = ["md", "org", "edn", "json", "txt", "csv"]

IGNORE_PADRAO = {
    "prefixes": [],
    "exact": [],
    "dirNames": ["node_modules", ".git"],
    "suffixes": [".DS_Store"],
    "hidden": True,  # qualquer componente começando com "." (mas não "..")
}

DIR_RECICLAGEM_PADRAO = ".recycle"
MAX_BYTES_DE_CONTEUDO_PADRAO = 16 * 1024 * 1024


def _ext(p):
    _, _, ext = p.rpartition(".")
    return ext.lower() if "." in p else ""


def _tem_componente_oculto(rel):
    return any(parte.startswith(".") and parte != ".." for parte in rel.split("/") if parte)


def _fazer_ignorador(ignore):
    cfg = dict(IGNORE_PADRAO)
    cfg.update(ignore or {})

    def ignorado(rel):
        p = rel.replace(os.sep, "/")
        partes = [x for x in p.split("/") if x]
        if any(p.startswith(x) for x in cfg["prefixes"]):
            return True
        if p in cfg["exact"]:
            return True
        if any(parte in cfg["dirNames"] for parte in partes):
            return True
        if any(p.endswith(x) for x in cfg["suffixes"]):
            return True
        return False

    ignorado.cfg = cfg
    return ignorado


def _info(st, caminho):
    return {
        "type": "directory" if os.path.isdir(caminho) else "file",
        "size": st.st_size,
        # Epoch em MILISSEGUNDOS, como o resto do contrato. Segundos aqui fariam toda comparação de
        # mtime com o frontend errar por três ordens de grandeza — e o sintoma seria "o arquivo
        # mudou" respondido sempre, ou nunca.
        "mtime": int(st.st_mtime * 1000),
    }


def criar_fs_do_app(root, content_extensions=None, ignore=None, recycle_dir=None,
                    max_content_bytes=None, ao_avisar=None):
    """Monta as operações confinadas a `root`."""
    raiz_real = resolver_raiz(root)
    extensoes = set(x.lower() for x in (content_extensions or EXTENSOES_DE_CONTEUDO_PADRAO))
    ignorado = _fazer_ignorador(ignore)
    ocultos_ignorados = ignorado.cfg["hidden"]
    dir_reciclagem = recycle_dir or DIR_RECICLAGEM_PADRAO
    max_conteudo = MAX_BYTES_DE_CONTEUDO_PADRAO if max_content_bytes is None else max_content_bytes
    avisar = ao_avisar or (lambda _e: None)

    def resolver(entrada):
        return resolver_na_raiz(raiz_real, entrada)

    def caminhar(inicio, aplicar_ignore=False):
        """Recursivo: descarta symlinks (um link para fora não vaza nem por leitura), devolve só
        arquivos, caminhos absolutos.

        Ocultos NÃO são descartados incondicionalmente — quem decide é sempre a config, senão
        `hidden: False` mentiria: o caminhamento já teria derrubado os ocultos antes da regra rodar.

        `aplicar_ignore` é opt-in porque as duas chamadoras querem coisas diferentes, e isso é
        deliberado: quem coleta conteúdo quer a visão filtrada, e quem lista quer a crua — filtrar
        lá esconderia do consumidor exatamente o que ele foi buscar.
        """
        saida = []
        fila = [inicio]
        while fila:
            diretorio = fila.pop()
            try:
                entradas = list(os.scandir(diretorio))
            except FileNotFoundError:
                continue
            for e in entradas:
                if e.is_symlink():
                    continue
                absoluto = os.path.join(diretorio, e.name)
                rel = os.path.relpath(absoluto, raiz_real).replace(os.sep, "/")
                if ocultos_ignorados and _tem_componente_oculto(rel):
                    continue
                if aplicar_ignore and ignorado(rel):
                    continue
                if e.is_dir():
                    fila.append(absoluto)
                else:
                    saida.append(absoluto)
        saida.sort()
        return saida

    def coletar(dir_abs):
        arquivos = []
        for absoluto in caminhar(dir_abs, aplicar_ignore=True):
            if _ext(absoluto) not in extensoes:
                continue
            try:
                st = os.stat(absoluto)
            except FileNotFoundError:
                continue  # corrida com escrita externa: só ignora
            if st.st_size > max_conteudo:
                # OMITIDO, e não listado com conteúdo vazio: um arquivo que o app não conhece fica
                # intocado, enquanto um listado como vazio é um arquivo vazio aos olhos dele — e o
                # primeiro save por cima apaga o conteúdo real no disco.
                avisar({"event": "file-skipped", "reason": "too-large", "path": absoluto,
                        "size": st.st_size, "limit": max_conteudo})
                continue
            with open(absoluto, "r", encoding="utf-8", errors="replace") as fh:
                conteudo = fh.read()
            info = _info(st, absoluto)
            arquivos.append({"path": absoluto, "content": conteudo, "size": info["size"],
                             "mtime": info["mtime"], "type": info["type"]})
        return arquivos

    class _Fs:
        root = raiz_real

        def stat(self, entrada):
            absoluto, _ = resolver(entrada)
            try:
                return _info(os.stat(absoluto), absoluto)
            except FileNotFoundError:
                raise nao_encontrado(entrada) from None

        def exists(self, entrada):
            """"Existe?" é pergunta booleana e merece resposta booleana.

            `stat` responde com erro quando não existe, e isso está certo — é a resposta correta
            para "me dê os metadados". Mas quem só quer saber se existe acaba usando o erro como
            fluxo de controle, e aí toda sondagem de rotina vira ruído de erro nos dois lados.
            Caminho FORA da raiz continua sendo EACCES: se existe ou não lá fora não é resposta que
            esta lib deva dar.
            """
            absoluto, _ = resolver(entrada)
            return {"exists": os.path.exists(absoluto)}

        def read_file(self, entrada):
            absoluto, _ = resolver(entrada)
            try:
                with open(absoluto, "r", encoding="utf-8", errors="replace") as fh:
                    return {"content": fh.read()}
            except FileNotFoundError:
                raise nao_encontrado(entrada) from None
            except IsADirectoryError:
                raise ErroFs(EINVAL, f"não é arquivo: {absoluto}") from None

        def write_file(self, entrada, conteudo):
            absoluto, _ = resolver(entrada)

            # Gravar por cima de um diretório nunca é o que o chamador quis: significa que o
            # caminho chegou errado. Sem esta checagem o open lança IsADirectoryError, que o
            # transporte não sabe classificar e devolve 500 — um pedido inválido do cliente
            # aparecendo como defeito do servidor.
            if os.path.isdir(absoluto):
                raise ErroFs(EINVAL, f"destino de escrita é um diretório: {absoluto}")
            if not isinstance(conteudo, (str, bytes, bytearray, memoryview)):
                raise ErroFs(EINVAL, f"conteúdo deve ser texto ou bytes, veio {type(conteudo).__name__}")

            os.makedirs(os.path.dirname(absoluto), exist_ok=True)
            if isinstance(conteudo, str):
                with open(absoluto, "w", encoding="utf-8") as fh:
                    fh.write(conteudo)
            else:
                with open(absoluto, "wb") as fh:
                    fh.write(bytes(conteudo))
            info = _info(os.stat(absoluto), absoluto)
            return {"path": absoluto, "size": info["size"], "mtime": info["mtime"]}

        def mkdir(self, entrada):
            absoluto, _ = resolver(entrada)
            try:
                os.mkdir(absoluto)
            except FileExistsError:
                pass  # diretório já existente não é erro — é o estado pedido
            return {}

        def mkdir_recur(self, entrada):
            absoluto, _ = resolver(entrada)
            os.makedirs(absoluto, exist_ok=True)
            return {}

        def readdir(self, entrada=None, aplicar_ignore=False):
            absoluto, _ = resolver(entrada or raiz_real)
            return caminhar(absoluto, aplicar_ignore=aplicar_ignore)

        def unlink(self, entrada, reciclar=True):
            """Por padrão NÃO apaga: move para a reciclagem com o caminho relativo achatado.

            `reciclar=False` apaga de verdade — para app que já tem lixeira própria, ou para o que
            é descartável por natureza (cache) e não deve inchar o diretório de reciclagem.
            """
            absoluto, rel = resolver(entrada)
            if not reciclar:
                try:
                    os.unlink(absoluto)
                except FileNotFoundError:
                    raise nao_encontrado(entrada) from None
                return {"recycled": None}

            reciclagem, _ = resolver(dir_reciclagem)
            os.makedirs(reciclagem, exist_ok=True)
            achatado = rel.replace(os.sep, "/").replace("/", "_")
            destino = os.path.join(reciclagem, achatado)
            try:
                os.replace(absoluto, destino)
            except FileNotFoundError:
                raise nao_encontrado(entrada) from None
            except OSError as err:
                if getattr(err, "errno", None) == 18:  # EXDEV: outro device (bind mount)
                    shutil.copyfile(absoluto, destino)
                    os.unlink(absoluto)
                else:
                    raise
            return {"recycled": destino}

        def rename(self, de, para):
            de_abs, _ = resolver(de)
            para_abs, _ = resolver(para)
            os.makedirs(os.path.dirname(para_abs), exist_ok=True)
            try:
                os.replace(de_abs, para_abs)
            except FileNotFoundError:
                raise nao_encontrado(de) from None
            return {}

        def copy(self, de, para):
            """Sobrescreve sem confirmação — o contrato de wire pede isso explicitamente."""
            de_abs, _ = resolver(de)
            para_abs, _ = resolver(para)
            os.makedirs(os.path.dirname(para_abs), exist_ok=True)
            try:
                shutil.copyfile(de_abs, para_abs)
            except FileNotFoundError:
                raise nao_encontrado(de) from None
            return {}

        def open_dir(self, entrada=None):
            absoluto, _ = resolver(entrada or raiz_real)
            if not os.path.exists(absoluto):
                raise nao_encontrado(entrada or raiz_real)
            if not os.path.isdir(absoluto):
                raise ErroFs(EINVAL, f"não é diretório: {absoluto}")
            return {"path": absoluto, "files": coletar(absoluto)}

        def get_files(self, entrada=None):
            return self.open_dir(entrada)

        def open_read(self, entrada):
            """Metadados + um abridor, para o transporte servir binário sem passar por JSON."""
            absoluto, _ = resolver(entrada)
            if not os.path.exists(absoluto) or os.path.isdir(absoluto):
                raise nao_encontrado(entrada)
            info = _info(os.stat(absoluto), absoluto)
            return {"path": absoluto, "size": info["size"], "mtime": info["mtime"],
                    "abrir": lambda: open(absoluto, "rb")}

    return _Fs()
